### PR TITLE
fix: move websocket config under config.route.v3.RouteAction

### LINKED ISSUE #2632: Move web socket upgrade config to route
*Description*:
>  Secure websocket is not working on envoy gateway, If I connect to the server without  tls, just plain ws://, everything works fine however once I connect with wss:// of which the connection is terminated at the lb as shown in the deployment file below, the connection fails. My assumption was initially a simple mis configuration I however haven't managed to get it working.

*Repro steps*:
>  
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: websocket-test
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: websocket-test
  template:
    metadata:
      labels:
        app.kubernetes.io/name: websocket-test
    spec:
      containers:
        - name: websocket-test
          image: jmalloc/echo-server           
          imagePullPolicy: IfNotPresent
          resources:
            limits:
              cpu: 300m
              memory: 500Mi
            requests:
              cpu: 10m
              memory: 80Mi
          ports:
            - name: default
              containerPort: 80
          env:
            - name: PORT
              value: '80'

---
apiVersion: v1
kind: Service
metadata:
  name: websocket-test
  labels:
    app.kubernetes.io/name: websocket-test
spec:
  ports:
  - name: default
    port: 80
    targetPort: 80
  selector:
    app.kubernetes.io/name: websocket-test

---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: websocket-test
spec:
  parentRefs:
    - kind: Gateway
      name: default
      namespace: envoy-gateway-system
  hostnames:
    - "test.domain.com"
  rules:
    - matches:
        - path:
            value: "/"
      backendRefs:
        - group: ""
          kind: Service
          name: websocket-test
          port: 80
          weight: 1

---

apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: default
  namespace: envoy-gateway-system
spec:
  gatewayClassName: envoy-gateway
  listeners:
    - name: http
      protocol: HTTP
      port: 80
      allowedRoutes:
        namespaces:
          from: Selector
          selector:
            matchLabels:
              gateway-enabled: "true"
    - name: https
      protocol: HTTPS
      port: 443
      allowedRoutes:
        namespaces:
          from: Selector
          selector:
            matchLabels:
              gateway-enabled: "true"
      tls:
        mode: Terminate
        certificateRefs:
          - name: cloudflare-origin-cert

---
```

*Environment*:
> - Gateway version is : docker.io/envoyproxy/gateway:v0.6.0 
   - Envoy version is : envoyproxy/envoy:distroless-v1.28-latest
   
*Logs*:
```console
envoy-envoy-gateway-system-default-b435bbfd-58c8b4d8d-j4r6h {"start_time":"2024-02-16T13:25:57.423Z","method":"GET","x-envoy-origin-path":"/","protocol":"HTTP/1.1","response_code":"403","response_flags":"-","response_code_details":"upgrade_failed","connection_termination_details":"-","upstream_transport_failure_reason":"-","bytes_received":"0","bytes_sent":"0","duration":"0","x-envoy-upstream-service-time":"-","x-forwarded-for":"197.239.12.79,10.244.1.69","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36","x-request-id":"e041de6c-b9d9-43d7-8395-ed023cb92771",":authority":"ws.stawi.io","upstream_host":"-","upstream_cluster":"httproute/realtime/websocket-test/rule/0","upstream_local_address":"-","downstream_local_address":"10.244.5.222:10443","downstream_remote_address":"10.244.1.69:38732","requested_server_name":"-","route_name":"httproute/realtime/websocket-test/rule/0/match/0/ws_stawi_io"}                                                                                                                                                                                                                                                                           
envoy-envoy-gateway-system-default-b435bbfd-58c8b4d8d-cxkrc {"start_time":"2024-02-16T13:26:03.477Z","method":"GET","x-envoy-origin-path":"/","protocol":"HTTP/1.1","response_code":"403","response_flags":"-","response_code_details":"upgrade_failed","connection_termination_details":"-","upstream_transport_failure_reason":"-","bytes_received":"0","bytes_sent":"0","duration":"0","x-envoy-upstream-service-time":"-","x-forwarded-for":"197.239.12.79,10.244.1.69","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36","x-request-id":"d0ca06e8-0872-4f94-9c41-04a2c3ee1fc9",":authority":"ws.stawi.io","upstream_host":"-","upstream_cluster":"httproute/realtime/websocket-test/rule/0","upstream_local_address":"-","downstream_local_address":"10.244.4.70:10443","downstream_remote_address":"10.244.1.69:46876","requested_server_name":"-","route_name":"httproute/realtime/websocket-test/rule/0/match/0/ws_stawi_io"}                                                                                                                                                                                                                                                                            
envoy-envoy-gateway-system-default-b435bbfd-58c8b4d8d-j4r6h {"start_time":"2024-02-16T13:26:09.396Z","method":"GET","x-envoy-origin-path":"/","protocol":"HTTP/1.1","response_code":"403","response_flags":"-","response_code_details":"upgrade_failed","connection_termination_details":"-","upstream_transport_failure_reason":"-","bytes_received":"0","bytes_sent":"0","duration":"0","x-envoy-upstream-service-time":"-","x-forwarded-for":"197.239.12.79,10.244.1.69","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36","x-request-id":"cf65f33f-216e-410d-b62e-0017731e34db",":authority":"ws.stawi.io","upstream_host":"-","upstream_cluster":"httproute/realtime/websocket-test/rule/0","upstream_local_address":"-","downstream_local_address":"10.244.5.222:10443","downstream_remote_address":"10.244.1.69:37792","requested_server_name":"-","route_name":"httproute/realtime/websocket-test/rule/0/match/0/ws_stawi_io"}
```

### PR DESCRIPTION
**Pull request by zetaab**
**What type of PR is this?**

fix: move websocket config under config.route.v3.RouteAction

**What this PR does / why we need it**:

If there were 1 grpc route under same listener, the websocket connections did not work at all. This should make websockets working under same listener which contains grpc routes.

**Which issue(s) this PR fixes**:

Fixes #2632

cc @arkodg

### CODE REVIEW

File: internal/cmd/egctl/testdata/translate/out/default-resources.all.yaml [hunk 1]
```diff
@@ -1041,3 +1041,5 @@ xds:
               name: grpcroute/default/backend/rule/0/match/0/www_grpc-example_com
               route:
                 cluster: grpcroute/default/backend/rule/0
+                upgradeConfigs:
```
Comment (zetaab) [inline internal/cmd/egctl/testdata/translate/out/default-resources.all.yaml:1044]: humm? why it does add this to grpcroute. I think it should not do that

File: internal/cmd/egctl/testdata/translate/out/from-gateway-api-to-xds.all.yaml [hunk 2]
```diff
@@ -567,3 +567,5 @@ xds:
               name: grpcroute/default/backend/rule/0/match/0/www_grpc-example_com
               route:
                 cluster: grpcroute/default/backend/rule/0
+                upgradeConfigs:
```
Comment (zetaab) [inline internal/cmd/egctl/testdata/translate/out/from-gateway-api-to-xds.all.yaml:570]: here as well adding to grpcroute?

File: internal/cmd/egctl/testdata/translate/out/from-gateway-api-to-xds.route.yaml [hunk 3]
```diff
@@ -30,3 +32,5 @@ xds:
             name: grpcroute/default/backend/rule/0/match/0/www_grpc-example_com
             route:
               cluster: grpcroute/default/backend/rule/0
+              upgradeConfigs:
```
Comment (zetaab) [inline internal/cmd/egctl/testdata/translate/out/from-gateway-api-to-xds.route.yaml:35]: and here

File: internal/xds/translator/route.go [hunk 4]
```diff
@@ -57,20 +57,25 @@ func buildXdsRoute(httpRoute *ir.HTTPRoute) (*routev3.Route, error) {
...
-			routeAction := buildXdsRouteAction(httpRoute.Destination.Name)
+			routeAction = buildXdsWeightedRouteAction(httpRoute)
 			if httpRoute.Mirrors != nil {
 				routeAction.RequestMirrorPolicies = buildXdsRequestMirrorPolicies(httpRoute.Mirrors)
 			}
-			router.Action = &routev3.Route_Route{Route: routeAction}
 		}
+		// Allow websocket upgrades for HTTP 1.1
+		// Reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism
+		routeAction.UpgradeConfigs = []*routev3.RouteAction_UpgradeConfig{
```
Comment (zetaab) [inline internal/xds/translator/route.go:73]: I do not understand why it is modifying grpc routes as this the only place where websocket configuration is modified. This function should only handle httpRoutes?

**Comment by zetaab** [comment]
looks like grpc stuff is also executed through that function. There is no way to see is the httpRoute grpc or not, one possible way is to check that from the name, but not perfect way do it. Maybe that should be stored to some variable earlier

File: internal/xds/translator/route.go [hunk 5]
```diff
@@ -57,20 +57,27 @@ func buildXdsRoute(httpRoute *ir.HTTPRoute) (*routev3.Route, error) {
 
 		router.Action = &routev3.Route_Route{Route: routeAction}
 	default:
+		routeAction := buildXdsRouteAction(httpRoute.Destination.Name)
```
Comment (arkodg) [inline internal/xds/translator/route.go:60]: thoughts on 
```
var routeAction *<type>
if httpRoute.BackendWeights.Invalid != 0 {
  routeAction = buildXdsWeightedRouteAction(httpRoute)
else {
  routeAction = buildXdsRouteAction(httpRoute.Destination.Name)
}  

if httpRoute.Mirrors != nil {
  routeAction.RequestMirrorPolicies = buildXdsRequestMirrorPolicies(httpRoute.Mirrors)
}
```

**Review by shawnh2 — APPROVED** [review]

**Pull request merged**